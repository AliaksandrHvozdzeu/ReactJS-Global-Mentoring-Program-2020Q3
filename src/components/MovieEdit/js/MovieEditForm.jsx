import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import moment from 'moment';
import CloseButton from '../../CloseButton';
import Text from '../../Text';
import Input from '../../Input';
import Calendar from '../../Calendar';
import MultiSelect from '../../MultiSelectInput';
import ResetButton from '../../ResetButton';
import Constants from '../../../constants';

const MovieEditForm = (
  {
    closeAction,
    modalTitle,
    initialState = {},
    availableGenres,
    onSubmit,
  },
) => {

  const onSave = useCallback((data) => {
    onSubmit(data);
    closeAction();
  }, [onSubmit]);

  const schema = Yup.object().shape({
    title: Yup.string().required(Constants.FORMIK_ERRORS.title),
    poster_path: Yup.string().url().required(Constants.FORMIK_ERRORS.poster_path),
    overview: Yup.string().required(Constants.FORMIK_ERRORS.overview),
    runtime: Yup.number()
      .min(0)
      .required(Constants.FORMIK_ERRORS.runtime),
    release_date: Yup.string()
      .transform((value) => {
        const dateValue = moment(value);
        return dateValue.isValid() ? dateValue.format(Constants.FORMIK_ERRORS.DEFAULT_DATE_FORMAT) : '';
      })
      .required(Constants.FORMIK_ERRORS.release_date),
    genres: Yup.array().min(1).required(Constants.FORMIK_ERRORS.genre),
  });

  const formik = useFormik({
    initialValues: initialState,
    validationSchema: schema,
    onSubmit: onSave,
  });

  const onSelectStateChange = useCallback(
    (values) => {
      const genres = values.map((item) => item.value);
      formik.setFieldValue('genres', genres, true);
    },
    [formik],
  );

  const getErrorMessageByField = useCallback(
    (fieldName) => {
      if (formik.touched[fieldName] && formik.errors[fieldName]) {
        return formik.errors[fieldName];
      }
      return '';
    },
    [formik],
  );

  return (
    <>
      <CloseButton closeAction={closeAction}/>
      <form className="add-movie" onSubmit={formik.handleSubmit}>
        <p className="modal-title">{modalTitle}</p>
        <Text title="Movie id"
              value={formik.values.id}
              name="id"
              id="movie-id"/>
        <Input id="movie-title"
               title="Title"
               name="title"
               type="text"
               visibility=""
               value={formik.values.title}
               onChange={formik.handleChange}
               onBlur={formik.handleBlur}
               error={getErrorMessageByField('title')}/>
        <Calendar id="movie-release"
                  title="Release date"
                  name="release_date"
                  type="date"
                  value={formik.values.release_date}
                  onChange={formik.handleChange}
                  onBlur={formik.handleBlur}
                  error={getErrorMessageByField('release_date')}/>
        <Input id="movie-url"
               title="Movie url"
               visibility=""
               name="poster_path"
               type="url"
               value={formik.values.poster_path}
               onChange={formik.handleChange}
               onBlur={formik.handleBlur}
               error={getErrorMessageByField('poster_path')}/>
        <MultiSelect title="genre"
                     options={availableGenres}
                     onAction={onSelectStateChange}
                     preselected={formik.values.genres}
                     onBlur={formik.handleBlur}
                     error={getErrorMessageByField('genre')}/>
        <Input id="movie-overview"
               title="Overview"
               visibility=""
               type="text"
               name="overview"
               value={formik.values.overview}
               onChange={formik.handleChange}
               onBlur={formik.handleBlur}
               error={getErrorMessageByField('overview')}/>
        <Input id="movie-runtime"
               title="Runtime"
               visibility=""
               name="runtime"
               type="number"
               value={formik.values.runtime}
               onChange={formik.handleChange}
               onBlur={formik.handleBlur}
               error={getErrorMessageByField('runtime')}/>
        <div className="add-movie-actions">
          <ResetButton onReset={formik.resetForm}/>
          <button type="button"
                  className="submit-button"
                  onClick={formik.handleSubmit}>SAVE
          </button>
        </div>
      </form>
    </>
  );

};

MovieEditForm.propTypes = {
  closeAction: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  modalTitle: PropTypes.string.isRequired,
  initialState: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    title: PropTypes.string,
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    overview: PropTypes.string,
    runtime: PropTypes.number,
  }).isRequired,
  availableGenres: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default MovieEditForm;
