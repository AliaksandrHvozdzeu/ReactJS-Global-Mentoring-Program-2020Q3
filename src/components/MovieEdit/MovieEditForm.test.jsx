import React from "react";
import rerender from "react-test-renderer";
import MovieEditForm from "./js/MovieEditForm";
import { jest } from "@jest/globals";

describe("when MovieEditForm", () => {

  test("then snapshot created", () => {
    const closeAction = jest.fn();
    const modalTitle = "EDIT FORM";
    const initialState = {};
    const onSubmit = jest.fn();
    const component = rerender.create(<MovieEditForm closeAction={closeAction}
                                                     modalTitle={modalTitle}
                                                     initialState={initialState}
                                                     availableGenres={[]}
                                                     onSubmit={onSubmit}/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
