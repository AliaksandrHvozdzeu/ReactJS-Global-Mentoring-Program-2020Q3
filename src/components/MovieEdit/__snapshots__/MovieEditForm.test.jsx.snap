// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when MovieEditForm then snapshot created 1`] = `
Array [
  <button
    aria-label="close"
    className="close-button"
    onClick={[MockFunction]}
    type="button"
  />,
  <form
    className="add-movie"
    onSubmit={[Function]}
  >
    <p
      className="modal-title"
    >
      EDIT FORM
    </p>
    <label
      className=""
      htmlFor="movie-title"
    >
      Title
      <input
        className="input"
        id="movie-title"
        name="title"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="text"
      />
    </label>
    <label
      htmlFor="movie-release"
    >
      Release date
      <input
        className="input"
        id="movie-release"
        name="release_date"
        onBlur={[Function]}
        onChange={[Function]}
        type="date"
        value=""
      />
    </label>
    <label
      className=""
      htmlFor="movie-url"
    >
      Movie url
      <input
        className="input"
        id="movie-url"
        name="poster_path"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="url"
      />
    </label>
    <label
      className="genres"
      htmlFor="genres"
      id="malty"
    >
      genre
      <div
        className="go2139093995 multi-select"
      >
        <div
          aria-expanded={false}
          aria-labelledby="malty"
          aria-readonly={true}
          className="go2646822163 dropdown-container"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={0}
        >
          <div
            className="go2642161244 dropdown-heading"
            onClick={[Function]}
          >
            <div
              className="dropdown-heading-value"
            >
              <span
                className="gray"
              >
                Select...
              </span>
            </div>
            <img
              alt="arrow_down"
              className="arrow"
              src="/images/arrow_down.png"
            />
          </div>
        </div>
      </div>
    </label>
    <label
      className=""
      htmlFor="movie-overview"
    >
      Overview
      <input
        className="input"
        id="movie-overview"
        name="overview"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="text"
      />
    </label>
    <label
      className=""
      htmlFor="movie-runtime"
    >
      Runtime
      <input
        className="input"
        id="movie-runtime"
        name="runtime"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="number"
      />
    </label>
    <div
      className="add-movie-actions"
    >
      <button
        className="reset-button"
        onClick={[Function]}
        type="button"
      >
        RESET
      </button>
      <button
        className="submit-button"
        onClick={[Function]}
        type="button"
      >
        SAVE
      </button>
    </div>
  </form>,
]
`;
